File: rabt-mobile/lib/models/advert.dart
// coverage:ignore-file
import 'package:freezed_annotation/freezed_annotation.dart';
import 'enums.dart';
import 'skill.dart';
import 'organizer.dart';

part 'advert.freezed.dart';
part 'advert.g.dart';

@freezed
class RecurringDays with _$RecurringDays {
  @JsonSerializable(fieldRename: FieldRename.snake)
  const factory RecurringDays({
    required String day, // backend uses string day key
    required List<DayTimePeriod> periods,
  }) = _RecurringDays;

  factory RecurringDays.fromJson(Map<String, dynamic> json) => _$RecurringDaysFromJson(json);
}

@freezed
class RecurringAdvertDetails with _$RecurringAdvertDetails {
  @JsonSerializable(fieldRename: FieldRename.snake)
  const factory RecurringAdvertDetails({
    required RecurrenceType recurrence,
    required TimeCommitment timeCommitmentPerSession,
    required DurationType duration,
    required List<RecurringDays> specificDays,
  }) = _RecurringAdvertDetails;

  factory RecurringAdvertDetails.fromJson(Map<String, dynamic> json) => _$RecurringAdvertDetailsFromJson(json);
}

@freezed
class OneOffAdvertDetails with _$OneOffAdvertDetails {
  @JsonSerializable(fieldRename: FieldRename.snake)
  const factory OneOffAdvertDetails({
    required DateTime eventDatetime,
    required TimeCommitment timeCommitment,
    required DateTime applicationDeadline,
  }) = _OneOffAdvertDetails;

  factory OneOffAdvertDetails.fromJson(Map<String, dynamic> json) => _$OneOffAdvertDetailsFromJson(json);
}

@freezed
class Advert with _$Advert {
  @JsonSerializable(fieldRename: FieldRename.snake)
  const factory Advert({
    required int id,
    required String title,
    required String description,
    required String category,
    required FrequencyType frequency,
    required int numberOfVolunteers,
    required LocationType locationType,
    String? addressText,
    String? postcode,
    String? imageUrl,
    String? city,
    required bool isActive,
    required OrganizerProfile organizer,
    @Default([]) List<Skill> requiredSkills,
    OneOffAdvertDetails? oneoffDetails,
    RecurringAdvertDetails? recurringDetails,
    required DateTime createdAt,
  }) = _Advert;

  factory Advert.fromJson(Map<String, dynamic> json) => _$AdvertFromJson(json);
}

File: rabt-backend/app/schemas/advert.py
from pydantic import BaseModel, Field, computed_field
from typing import Optional, List
from datetime import datetime
from app.database.enums import (
    FrequencyType,
    LocationType,
    RecurrenceType,
    TimeCommitment,
    DurationType,
    DayPeriod,
)
from app.schemas.skill import SkillResponse
from app.schemas.organizer import OrganizerResponse
from app.utils.file_utils import get_image_url


class RecurringDays(BaseModel):
    day: str
    periods: List[DayPeriod]


class RecurringAdvertDetails(BaseModel):
    recurrence: RecurrenceType
    time_commitment_per_session: TimeCommitment
    duration: DurationType
    specific_days: List[RecurringDays]


class OneOffAdvertDetails(BaseModel):
    event_datetime: datetime
    time_commitment: TimeCommitment
    application_deadline: datetime


class AdvertBase(BaseModel):
    title: str
    description: str
    category: str
    frequency: FrequencyType
    number_of_volunteers: int = 1
    location_type: LocationType
    address_text: Optional[str] = None
    postcode: Optional[str] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    advert_image_url: Optional[str] = None


class AdvertCreate(AdvertBase):
    required_skill_ids: List[int] = []
    oneoff_details: Optional[OneOffAdvertDetails] = None
    recurring_details: Optional[RecurringAdvertDetails] = None


class AdvertUpdate(AdvertBase):
    required_skill_ids: Optional[List[int]] = None
    oneoff_details: Optional[OneOffAdvertDetails] = None
    recurring_details: Optional[RecurringAdvertDetails] = None


class AdvertResponse(AdvertBase):
    id: int
    is_active: bool
    organizer: OrganizerResponse
    required_skills: List[SkillResponse] = []
    oneoff_details: Optional[OneOffAdvertDetails] = None
    recurring_details: Optional[RecurringAdvertDetails] = None
    created_at: datetime

    @computed_field
    @property
    def image_url(self) -> Optional[str]:
        """Convert relative image path to full URL"""
        return get_image_url(self.advert_image_url)

    class Config:
        from_attributes = True


class AdvertResponseForApplications(AdvertBase):
    id: int
    is_active: bool
    organizer: OrganizerResponse
    required_skills: List[SkillResponse] = []
    created_at: datetime

    @computed_field
    @property
    def image_url(self) -> Optional[str]:
        """Convert relative image path to full URL"""
        return get_image_url(self.advert_image_url)

    class Config:
        from_attributes = True


class AdvertListResponse(BaseModel):
    items: List[AdvertResponse]
    total_pages: int

class OrganizerAdvertsListResponse(BaseModel):
    items: List[AdvertResponse]
